noir-projects/aztec-nr/aztec/src/context/call_interfaces.nr

```diff
@@ -0,0 +1,407 @@
+        let returns = context.call_public_function_with_packed_args(self.target_contract, self.selector, self.args_hash, true, false);
+        returns.deserialize_into()
+    }
+
+    pub fn enqueue_view(self, context: &mut PrivateContext) {
+        context.call_public_function_with_packed_args(self.target_contract, self.selector, self.args_hash, true, false)
+    }
+}
+
+struct PublicStaticVoidCallInterface {
+    target_contract: AztecAddress,
+    selector: FunctionSelector,
+    args_hash: Field
+}
+
+impl PublicStaticVoidCallInterface {
+    pub fn view(self, context: &mut PublicContext) {
+        context.call_public_function_with_packed_args(self.target_contract, self.selector, self.args_hash, true, false).assert_empty();
+    }
+
+    pub fn enqueue_view(self, context: &mut PrivateContext) {
+        context.call_public_function_with_packed_args(self.target_contract, self.selector, self.args_hash, true, false)
+    }
+}
+
+struct AvmCallInterface<T> {
+    target_contract: AztecAddress,
+    selector: FunctionSelector,
+    args: [Field],
+    gas_opts: GasOpts,
+}
+
+impl<T> AvmCallInterface<T> {
+    pub fn with_gas(self: &mut Self, gas_opts: GasOpts) -> &mut Self {
+        self.gas_opts = gas_opts;
+        self
+    }
+
+    pub fn call<N>(self, context: &mut AvmContext) -> T where T: Deserialize<N> {
+        let returns = context.call_public_function(self.target_contract, self.selector, self.args, self.gas_opts);
+        returns.deserialize_into()
+    }
+
+    pub fn view<N>(self, context: &mut AvmContext) -> T where T: Deserialize<N> {
+        let returns = context.static_call_public_function(self.target_contract, self.selector, self.args, self.gas_opts);
+        returns.deserialize_into()
+    }
+
+    pub fn delegate_call<N>(self, context: &mut AvmContext) -> T where T: Deserialize<N> {
+        let returns = context.delegate_call_public_function(self.target_contract, self.selector, self.args);
+        returns.deserialize_into()
+    }
+
+    pub fn enqueue(self, context: &mut PrivateContext) {
+        // This packing is only here because PrivateContext's call_public* functions do not accept a slice for the args.
+        let args_hash = arguments::pack_arguments(self.args);
+        context.call_public_function_with_packed_args(
+            self.target_contract,
+            self.selector,
+            args_hash,
+            /*static=*/ false,
+            /*delegate=*/ false
+        )
+    }
+
+    pub fn enqueue_view(self, context: &mut PrivateContext) {
+        // This packing is only here because PrivateContext's call_public* functions do not accept a slice for the args.
+        let args_hash = arguments::pack_arguments(self.args);
+        context.call_public_function_with_packed_args(
+            self.target_contract,
+            self.selector,
+            args_hash,
+            /*static=*/ true,
+            /*delegate=*/ false
```

Comment:
I'd consider adding similar comments to the other callsites, it's otherwise a bit annoying to follow